Directory revamp

dgobalak opened the pull request:
# Purpose
Change to the directory structure outlined here: https://www.notion.so/uworbital/FW-Next-Comp-Planning-f04acfffbc6c49fc86d8486fd48e70ab?pvs=4#026f8eb4d991459a8f8ff73fb6876d21 **Note: The Notion page might be out-of-date**

**Goal:** 
- organize the repo so we can put all projects (gs, OBC, etc) in one repo.
- update the build system to handle tests, OBC fw, and gs code. Our current build system doesn't use CMake to build the gs. 
- Decouple the obc_gs_interface from the OBC fw
- Split things up into more logical sections. I wanted to make sure we didn't have a situation like last comp where `cdh/` had different things (command manager, telemetry manager, timekeeper, supervisor) and we dumped a bunch of stuff into `common/` (some of which can be used by the ground station like packing utils and others are tied to the OBC fw like obc_time)

# New Changes
- Changed directory structure; I tried to minimize the number of folders at the top-level of the directory.
- Switched from include guards to `#pragma once` for a few reasons:
  1. It makes refactoring simpler
  2. It's less prone to bugs since you don't have to choose a unique include guard (that can sometimes be a lot of characters)
  3. It's less code to write
- Moved interface-related comms files to `obc_gs_interface` and added the `obc_gs_` prefix to each filename
- Added project-specific error codes. `obc_error_code_t` for OBC firmware, `obc_gs_error_code_t` for interface code, and `gs_error_code_t` for ground station code. Because of this, I've removed the use of `RETURN_IF_ERROR_CODE` from the interface for now. In a future PR, we'll add a logging module that'll work across the obc, gs, and interface.
- Created standalone obc_gs_interface lib that can be linked into any executable (obc, gs, or test)
- Removed any dependencies on the main firmware from the obc_gs_interface module (this allows us to compile it as a static lib)

Other stuff:
- Added telemetry unpack functions
- Added some unit tests (for telemetry pack and obc_time_utils)
- Moved ISR callbacks to `drivers/` to keep the HAL independent from the rest of our code. This should, in theory, allow us to compile the HAL as a static lib. I also added back the rtc alarm callback which was deleted accidentally in a past PR.

# Testing
- CI
- Tested fw on launchpad

# Outstanding Changes
- The CC1120 driver shouldn't be dependent on code from `services/`
- Make a static lib for the hal and then fix the builds for all the test apps. It'd be nice if we could just link the HAL static lib when building each test app.
- Create a logging module (with RETURN_IF_ERROR_CODE macros) that'll work with the different error codes we have
- Write CMake and unit test documentation
- Combine obc_state into supervisor and rethink what telemetry we send down.

manumanuk reviewed (changes requested):
- I don't know if I agree with switching to `#pragma once`, what if we need to use conditionals based on whether or not a certain header file preprocessor symbol is defined?
- I think maybe we should discuss a bit more about whether or not it makes sense to switch to this "interface", "obc", "eps" in one repo structure. I'm more and more convinced that it should be this way, but we should bring everyone into the conversation
- It's a bit confusing for the hal code to be both in the drivers and hal folders
- It's also a bit weird to have tests in both a test/ folder and a separate interface_tests folder
- Could you define a bit more what exactly would be in the "sys" folder?

dgobalak commented:
> * I don't know if I agree with switching to `#pragma once`, what if we need to use conditionals based on whether or not a certain header file preprocessor symbol is defined?
Idk when we'd have to do that, but if we do, then we can just #define something just for those files right?
> * I think maybe we should discuss a bit more about whether or not it makes sense to switch to this "interface", "obc", "eps" in one repo structure. I'm more and more convinced that it should be this way, but we should bring everyone into the conversation
Fair point. We could probably just talk about this during the next integration meeting
> * It's a bit confusing for the hal code to be both in the drivers and hal folders
The "hal" code in drivers is basically just the stuff that was in `drivers/common/` originally. I wasn't sure what to name it, so I just called it hal_wrappers. Not sure what the better alternative would be? 
> * It's also a bit weird to have tests in both a test/ folder and a separate interface_tests folder
Ya so I did that assuming we'd put interfaces in a separate repo and include it as a submodule. If we'd be using that submodule in the obc-firmware and ground station repos, we'd need a way to include the tests in the ground station repo as well.
> * Could you define a bit more what exactly would be in the "sys" folder?
Ya ngl, I just wanted to name it something other than common lol. But I couldn't think of a name. It basically just contains all the code (built on top of the drivers and hal) that's common to the different services. Lmk if u think there's a better name tho.

Gjjjiang reviewed (commented):
Just wondering the same thing as @manumanuk about the test folders. Shouldn't all tests be stored together in a separate test folder (maybe at least separate FW and GS tests) and in that folder, we can organize them further by interface and obc? 
Other than that, folder structure lgtm.

Gjjjiang commented:
> Just wondering the same thing as @manumanuk about the test folders. Shouldn't all tests be stored together in a separate test folder (maybe at least separate FW and GS tests) and in that folder, we can organize them further by interface and obc?
Oops, disregard, made a review right after you answered the question

dgobalak commented:
> I guess the question now is whether the interface going to be a separate repo?

Ya exactly. And I think that'll depend on if the ground station code is in a separate repo?

manumanuk reviewed (changes requested):
Some nitpicky items:
- Do we really need both obc/drivers/hal_wrappers and obc/drivers/peripherals? I'd say we bring everything inside drivers/peripherals/ outside, I think there's no need for that extra folder
- Rename sw_wdt driver to watchdog
- Does the config folder need to be separate?
- Can reliance edge be stored in libs?
- What is the point of bringing hal driver code out of the hal/ folder and into the drivers folder?- Thoughts on renaming services to modules?
- Why are there interface tests in obc/utils/bringup/interface_tests? Shouldn't we bring them out to the global tests folder?
- I feel like posix_freertos can be taken out of the utils folder and put into the global libs folder

In general, we should try to limit the number of folders and subfolders we have to maintain better simplicity

dgobalak commented:
> * Do we really need both obc/drivers/hal_wrappers and obc/drivers/peripherals? I'd say we bring everything inside drivers/peripherals/ outside, I think there's no need for that extra folder

Done

> * Rename sw_wdt driver to watchdog

Renamed to sw_watchdog (software watchdog) since there's also gonna be an external hardware watchdog.

> * Does the config folder need to be separate?

Moved task_config to services and board_config to drivers.

> * Can reliance edge be stored in libs?

I kinda wanted everything in libs to be standalone (i.e. we should be able to compile anything in libs as a static library or even object library would be ok). The problem with reliance edge is that it's dependent on FreeRTOS (i.e. our HAL), so it'd be weird to move it outside `obc/`. I've moved it outside `ext/` tho to reduce the number of directories.

> * What is the point of bringing hal driver code out of the hal/ folder and into the drivers folder?- Thoughts on renaming services to modules?

This is mainly so we can keep the HAL completely independent from the rest of the fw. Originally, because of notification.c, it was dependent on a few of our drivers.

> * Why are there interface tests in obc/utils/bringup/interface_tests? Shouldn't we bring them out to the global tests folder?

The "interface_tests" are just a tool we planned to use during board bringup. Their main purpose is debugging the various interfaces of the RM46. Although, we've never actually used the tool. Since these aren't really "tests", I'll just rename it to "interface_debug_utility" for now. We'll visit this again once we start looking into the HIL test setup since this _might_ be useful there.

> * I feel like posix_freertos can be taken out of the utils folder and put into the global libs folder

Done

> In general, we should try to limit the number of folders and subfolders we have to maintain better simplicity

Ya I agree. I just don't want a situation like we had with the `cdh` directory having a bunch of files for different tasks (command manager, telemetry manager, alarm handler, supervisor, etc.)

manumanuk commented:
Still think we should rename "sys" to something else (I kinda want to call it "utils" but there's so many other things called "utils" already) and "services" to "modules" or something

manumanuk commented:
Also, thoughts on renaming obc/drivers/common to obc/drivers/rm46?

dgobalak commented:
> Also, thoughts on renaming obc/drivers/common to obc/drivers/rm46?

Ya I like that

dgobalak commented:
> Still think we should rename "sys" to something else (I kinda want to call it "utils" but there's so many other things called "utils" already) and "services" to "modules" or something

Sure I can rename services. Idk about turning `sys` -> `utils` though, mainly cuz I think `utils` is the perfect name for the `sys/utils` subdirectory. I'd be ok with changing `sys` to `common` tho.

manumanuk reviewed (approved):
LGTM

Navtajh04 reviewed (approved):
LGTM :(
